README.rst
========
favicon
========

.. start-badges

.. image:: https://img.shields.io/pypi/v/favicon.svg
   :target: https://pypi.python.org/pypi/favicon

.. image:: https://img.shields.io/badge/license-MIT-blue.svg
   :target: /LICENSE

.. image:: https://travis-ci.org/scottwernervt/favicon.svg?branch=master
   :target: https://travis-ci.org/scottwernervt/favicon

.. end-badges

``favicon`` is a Python library to find a website's favicon.

Installation
============

.. code-block:: bash

   pip install favicon

Usage
=====

Get all icons:

.. code-block:: python

   >>> import favicon
   >>> icons = favicon.get('https://www.python.org/')
   Icon(url='https://www.python.org/static/apple-touch-icon-144x144-precomposed.png', width=144, height=144, format='png')
   Icon(url='https://www.python.org/static/apple-touch-icon-114x114-precomposed.png', width=114, height=114, format='png')
   Icon(url='https://www.python.org/static/apple-touch-icon-72x72-precomposed.png', width=72, height=72, format='png')
   Icon(url='https://www.python.org/static/apple-touch-icon-precomposed.png', width=0, height=0, format='png')
   Icon(url='https://www.python.org/static/favicon.ico', width=0, height=0, format='ico')

Download largest icon:

.. code-block:: python

   import requests
   import favicon

   icons = favicon.get('https://www.python.org/')
   icon = icons[0]

   response = requests.get(icon.url, stream=True)
   with open('/tmp/python-favicon.{}'.format(icon.format), 'wb') as image:
       for chunk in response.iter_content(1024):
           image.write(chunk)

   # /tmp/python-favicon.png

`Request library <https://2.python-requests.org/>`_ parameters can be passed to ``favicon.get()`` as keyword
arguments:

.. code-block:: python

   import favicon

   user_agent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/35.0.1916.47 Safari/537.36'
   headers = {'User-Agent': user_agent}
   favicon.get('https://www.python.org/', headers=headers, timeout=2)

Requirements
============

* `requests <http://docs.python-requests.org/>`_
* `beautifulsoup4 <https://www.crummy.com/software/BeautifulSoup/bs4/doc/>`_

Inspiration
===========

* `pyfav <https://github.com/phillipsm/pyfav>`_
* `besticon <https://github.com/mat/besticon/>`_
* `How to get high resolution website logo (favicon) for a given URL <https://stackoverflow.com/questions/21991044/how-to-get-high-resolution-website-logo-favicon-for-a-given-url>`_


CHANGELOG.rst
Changelog
=========

0.7.2 (2020-06-27)
------------------

* Added option for disabling verification by HTTP request of default /favicon.ico file, to have similar behavior
  to tag-based favicon results.
* Fixed flaky test.

0.7.1 (2020-06-17)
------------------

* Added support for receiving HTML input
* Fixed issue with dimensions(tag) not working for some websites

0.7.0 (2019-08-31)
------------------

* Handle empty ``href`` and ``content`` attribute values (`#22 <https://github.com/scottwernervt/favicon/issues/22>`_).
* Support passing request library parameters to ``favicon.get()`` (`#21 <https://github.com/scottwernervt/favicon/issues/21>`_).

  * Deprecate ``headers`` argument. Use keyword arguments: ``favicon.get(url, headers={'User-Agent'; 'my-agent'}``.

0.6.0 (2019-08-10)
------------------

* Upgrade ``beautifulsoup4`` and ``requests`` package dependencies.

0.5.1 (2018-11-05)
------------------

* Fix ``NoneType`` object has no attribute ``lower`` for meta tags (`#16 <https://github.com/scottwernervt/favicon/issues/16>`_).

0.5.0 (2018-11-05)
------------------

* Add support for meta tags (`#15 <https://github.com/scottwernervt/favicon/pull/15>`_).
* Set bs4 parser to ``html.parser`` (`#13 <https://github.com/scottwernervt/favicon/issues/13>`_).
* Use ``src`` package structure (`#11 <https://github.com/scottwernervt/favicon/issues/11>`_).

0.4.1 (2018-10-01)
------------------

* Update ``requirements.txt`` and ``dev-requirements.txt``.

0.4.0 (2018-07-19)
------------------

* Add support for Python 2.7 and PyPy.
* Get icon size for New York Times (`#9 <https://github.com/scottwernervt/favicon/issues/9>`_).

0.3.0 (2018-05-18)
------------------

* Fav icon not found for microsoft.com (`#7 <https://github.com/scottwernervt/favicon/issues/7>`_).

0.2.0 (2018-05-17)
------------------

* Handle poor html values in links (`#5 <https://github.com/scottwernervt/favicon/issues/5>`_).
* Use given website for icon url scheme (`#6 <https://github.com/scottwernervt/favicon/issues/6>`_).

0.1.0 (2018-05-07)
------------------

* First release.


tests/conftest.py
# -*- coding: utf-8 -*-
import pytest
import requests_mock


@pytest.fixture(scope='function')
def m():
    mocker = requests_mock.mock()
    mocker.head('http://mock.com/favicon.ico', text='404', status_code=404)

    mocker.start()
    yield mocker
    mocker.stop()


tests/test_favicon.py
# -*- coding: utf-8 -*-
import pytest
from bs4 import BeautifulSoup

import favicon
from favicon.favicon import is_absolute

s = BeautifulSoup(features='html.parser')


@pytest.mark.parametrize(
    'url,expected',
    [
        ('http://mock.com/', 'http://mock.com/favicon.ico'),
        ('https://mock.com/', 'https://mock.com/favicon.ico'),
        ('http://mock.com/mock/', 'http://mock.com/favicon.ico'),
        ('http://mock.com/mock/index.html', 'http://mock.com/favicon.ico'),
        (
            'http://mock.com/mock/index.html?q=mock',
            'http://mock.com/favicon.ico'
        ),
        (
            'http://mock.com:80/mock/index.html?q=mock',
            'http://mock.com:80/favicon.ico'
        ),
    ],
    ids=[
        'default',
        'https',
        'folder',
        'file',
        'parameter',
        'port',
    ],
)
def test_default(m, url, expected):
    m.get(url, text='body')
    m.head(expected, text='icon')
    m.get(expected, text='icon')

    icons = favicon.get(url)
    assert icons

    icon = icons[0]
    assert icon.url == expected


@pytest.mark.parametrize(
    'link',
    [
        '<link rel="icon" href="favicon.ico">',
        '<link rel="ICON" href="favicon.ico">',
        '<link rel="shortcut icon" href="favicon.ico">',
        '<link rel="apple-touch-icon" href="favicon.ico">',
        '<link rel="apple-touch-icon-precomposed" href="favicon.ico">',
    ],
    ids=[
        'icon',
        'ICON (#7)',
        'shortcut icon',
        'apple-touch-icon',
        'apple-touch-icon-precomposed',
    ],
)
def test_link_tag(m, link):
    m.get('http://mock.com/', text=link)

    icons = favicon.get('http://mock.com/')
    assert icons


@pytest.mark.parametrize(
    'link,size',
    [
        ('<link rel="icon" href="logo.png" sizes="any">', (0, 0)),
        ('<link rel="icon" href="logo.png" sizes="16x16">', (16, 16)),
        ('<link rel="icon" href="logo.png" sizes="24x24+">', (24, 24)),
        ('<link rel="icon" href="logo.png" sizes="32x32 64x64">', (64, 64)),
        ('<link rel="icon" href="logo.png" sizes="64x64 32x32">', (64, 64)),
        ('<link rel="icon" href="logo-128x128.png" sizes="any">', (128, 128)),
        (u'<link rel="icon" href="logo.png" sizes="16×16">', (16, 16)),
    ],
    ids=[
        'any',
        '16x16',
        '24x24+',
        '32x32 64x64',
        '64x64 32x32',
        'logo-128x128.png',
        'new york times (#9)',
    ],
)
def test_link_tag_sizes_attribute(m, link, size):
    m.get('http://mock.com/', text=link)

    icons = favicon.get('http://mock.com/')
    assert icons

    icon = icons[0]
    assert icon.width == size[0] and icon.height == size[1]


@pytest.mark.parametrize(
    'link,url',
    [
        ('<link rel="icon" href="logo.png">', 'http://mock.com/logo.png'),
        ('<link rel="icon" href="logo.png\t">', 'http://mock.com/logo.png'),
        (
            '<link rel="icon" href="/static/logo.png">',
            'http://mock.com/static/logo.png',
        ),
        (
            '<link rel="icon" href="https://cdn.mock.com/logo.png">',
            'https://cdn.mock.com/logo.png',
        ),
        (
            '<link rel="icon" href="//cdn.mock.com/logo.png">',
            'http://cdn.mock.com/logo.png',
        ),
        (
            '<link rel="icon" href="http://mock.com/logo.png?v2">',
            'http://mock.com/logo.png?v2',
        ),
    ],
    ids=[
        'filename',
        'filename \\t (#5)',
        'relative',
        'https',
        'forward slashes',
        'query string (#7)',
    ],
)
def test_link_tag_href_attribute(m, link, url):
    m.get('http://mock.com/', text=link)

    icons = favicon.get('http://mock.com/')
    assert icons

    icon = icons[0]
    assert icon.url == url


def test_link_tag_empty_href_attribute(m):
    """'NoneType' object has no attribute 'strip' #22"""
    m.get('http://mock.com/', text='<link rel="icon" href="">')

    with pytest.warns(None):
        icons = favicon.get('http://mock.com/')

    assert not icons


@pytest.mark.parametrize(
    'meta_tag',
    [
        '<meta name="msapplication-TileImage" content="favicon.ico">',
        '<meta name="msapplication-tileimage" content="favicon.ico">',
        '<meta property="og:image" content="favicon.png">',
    ],
    ids=['msapplication-TileImage', 'msapplication-tileimage', 'og:image'],
)
def test_meta_tag(m, meta_tag):
    m.get('http://mock.com/', text=meta_tag)

    icons = favicon.get('http://mock.com/')
    assert icons


def test_invalid_meta_tag(m):
    m.get(
        'http://mock.com/',
        text='<meta content="en-US" data-rh="true" itemprop="inLanguage"/>',
    )

    icons = favicon.get('http://mock.com/')
    assert not icons


def test_request_kwargs(m):
    """Add request timeout #21"""
    headers = {
        'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0'
    }
    m.get('http://mock.com/', request_headers=headers, text='body')

    favicon.get('http://mock.com/', headers=headers)

    # Test deprecated header argument
    with pytest.warns(DeprecationWarning):
        favicon.get('http://mock.com/', headers)


@pytest.mark.parametrize(
    'url,expected',
    [
        ('http://mock.com/favicon.ico', True),
        ('favicon.ico', False),
        ('/favicon.ico', False),
    ],
)
def test_is_absolute_helper(url, expected):
    assert is_absolute(url) == expected


def test_html_input():
    # contents of mock.com
    mock_com_html = '''<!DOCTYPE html><!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en-US" prefix="og: http://ogp.me/ns#"> <![endif]--><!--[if IE 7]><html class="no-js lt-ie9 lt-ie8" lang="en-US" prefix="og: http://ogp.me/ns#"> <![endif]--><!--[if IE 8]><html class="no-js lt-ie9" lang="en-US" prefix="og: http://ogp.me/ns#"> <![endif]--><!--[if gt IE 8]><!--><html class="no-js" lang="en-US" prefix="og: http://ogp.me/ns#"> <!--<![endif]--><head>    <meta charset="utf-8">    <title>Home - MOCK.com</title>    <meta name="viewport" content="width=device-width, initial-scale=1.0">    <link rel="shortcut icon" type="image/x-icon" href="http://mock.com/wp-content/uploads/2014/03/favicon.ico"/>    <link rel="canonical" href="http://mock.com/"/>    <meta property="og:locale" content="en_US"/>    <meta property="og:type" content="website"/>    <meta property="og:title" content="Home - MOCK.com"/>    <meta property="og:url" content="http://mock.com/"/>    <meta property="og:site_name" content="MOCK.com"/></head><body>Test</body></html>''' # noqa

    icons = favicon.get(
        'http://mock.com/',
        html_override=mock_com_html,
    )
    assert icons
    icons.sort(key=lambda icon: icon.url)
    assert icons[0].url == 'http://mock.com/favicon.ico'
    assert icons[1].url == 'http://mock.com/wp-content/uploads/2014/03/favicon.ico'


debug.py
"""

TODO: FEATURE: Check in manifest.json and browserconfig.xml for icons.

https://github.com/phillipsm/pyfav
https://github.com/phillipsm/pyfav/blob/master/pyfav/pyfav.py
https://stackoverflow.com/questions/21991044/how-to-get-high-resolution-website-logo-favicon-for-a-given-url
https://github.com/mat/besticon/

"""

import favicon

# urls = [
#     'http://www.autodesk.com',
#     'https://www.github.com',
#     'https://www.oracle.com',
#     'https://www.kicktipp.de/',
#     'https://www.bing.com',
#     'http://www.yelp.com',
#     'http://www.clearvoice.com',
#     'http://www.adeccousa.com',
#     'http://www.mindmatters.de',
#     'https://www.kogan.com',
# ]
#
# for url in urls:
#     icons = favicon.get(url)
#     print(url)
#     for icon in icons:
#         print(icon)
#     print()

icons = favicon.get('https://www.python.org/')
for icon in icons:
    print(icon)

src/favicon/favicon.py
# -*- coding: utf-8 -*-
"""favicon
:copyright: (c) 2019 by Scott Werner.
:license: MIT, see LICENSE for more details.
"""
import os
import re
import warnings

from collections import namedtuple

try:
    from urllib.parse import urljoin, urlparse, urlunparse
except ImportError:
    from urlparse import urljoin, urlparse, urlunparse

import requests

from bs4 import BeautifulSoup

__all__ = ['get', 'Icon']

HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_2) '
    'AppleWebKit/537.36 (KHTML, like Gecko) '
    'Chrome/33.0.1750.152 Safari/537.36'
}

LINK_RELS = [
    'icon',
    'shortcut icon',
    'apple-touch-icon',
    'apple-touch-icon-precomposed',
]

META_NAMES = ['msapplication-TileImage', 'og:image']

SIZE_RE = re.compile(r'(?P<width>\d{2,4})x(?P<height>\d{2,4})', flags=re.IGNORECASE)

Icon = namedtuple('Icon', ['url', 'width', 'height', 'format'])


def get(url, *args, html_override=None, verify_default_icon=True, **request_kwargs):
    """Get all fav icons for a url.

    :param url: Homepage.
    :type url: str

    :param html_override: HTML input, as string. Will be used instead of an HTTP response from
        the `url`.
    :type html_override: str or None

    :param verify_default_icon: Whether to verify the existence of the default
        https://www.domain.com/favicon.ico file, or to return it as a probable result which
        still requires verification, like the other results.
    :type verify_default_icon: bool

    :param request_kwargs: Request headers argument.
    :type request_kwargs: Dict

    :return: List of fav icons found sorted by icon dimension.
    :rtype: list[:class:`Icon`]
    """
    if args:  # backwards compatible with <= v0.6.0
        warnings.warn(
            "headers arg is deprecated. Use headers key in request_kwargs dict.",
            DeprecationWarning
        )
        request_kwargs.setdefault('headers', args[0])

    request_kwargs.setdefault('headers', HEADERS)
    request_kwargs.setdefault('allow_redirects', True)

    if html_override is None:
        response = requests.get(url, **request_kwargs)
        response.raise_for_status()
        final_url = response.url
        html_override = response.text
    else:
        final_url = url

    icons = set()

    default_icon = default(final_url, verify_default_icon, **request_kwargs)
    if default_icon:
        icons.add(default_icon)

    link_icons = tags(final_url, html_override)
    if link_icons:
        icons.update(link_icons)

    return sorted(icons, key=lambda i: i.width + i.height, reverse=True)


def default(url, verify_default_icon, **request_kwargs):
    """Get icon using default filename favicon.ico.

    :param url: Url for site.
    :type url: str

    :param verify_default_icon: Whether to verify the existence of the default
        https://www.domain.com/favicon.ico file, or to return it as a probable result
        which still requires verification, like the other results.
    :type verify_default_icon: bool

    :param request_kwargs: Request headers argument.
    :type request_kwargs: Dict

    :return: Icon or None.
    :rtype: :class:`Icon` or None
    """
    parsed = urlparse(url)
    favicon_url = urlunparse((parsed.scheme, parsed.netloc, 'favicon.ico', '', '', ''))
    if not verify_default_icon:
        return Icon(favicon_url, 0, 0, 'ico')

    response = requests.head(favicon_url, **request_kwargs)
    if response.status_code == 200:
        return Icon(response.url, 0, 0, 'ico')


def tags(url, html):
    """Get icons from link and meta tags.

    .. code-block:: html

       <link rel="apple-touch-icon" sizes="144x144" href="apple-touch-icon.png">
       <meta name="msapplication-TileImage" content="favicon.png">

    :param url: Url for site.
    :type url: str

    :param html: Body of homepage.
    :type html: str

    :return: Icons found.
    :rtype: set
    """
    soup = BeautifulSoup(html, features='html.parser')

    link_tags = set()
    for rel in LINK_RELS:
        for link_tag in soup.find_all(
            'link', attrs={'rel': lambda r: r and r.lower() == rel, 'href': True}
        ):
            link_tags.add(link_tag)

    meta_tags = set()
    for meta_tag in soup.find_all('meta', attrs={'content': True}):
        meta_type = meta_tag.get('name') or meta_tag.get('property') or ''
        meta_type = meta_type.lower()
        for name in META_NAMES:
            if meta_type == name.lower():
                meta_tags.add(meta_tag)

    icons = set()
    for tag in link_tags | meta_tags:
        href = tag.get('href', '') or tag.get('content', '')
        href = href.strip()

        if not href or href.startswith('data:image/'):
            continue

        if is_absolute(href):
            url_parsed = href
        else:
            url_parsed = urljoin(url, href)

        # repair '//cdn.network.com/favicon.png' or `icon.png?v2`
        scheme = urlparse(url).scheme
        url_parsed = urlparse(url_parsed, scheme=scheme)

        width, height = dimensions(tag)
        _, ext = os.path.splitext(url_parsed.path)

        icon = Icon(url_parsed.geturl(), width, height, ext[1:].lower())
        icons.add(icon)

    return icons


def is_absolute(url):
    """Check if url is absolute.

    :param url: Url for site.
    :type url: str

    :return: True if homepage and false if it has a path.
    :rtype: bool
    """
    return bool(urlparse(url).netloc)


def dimensions(tag):
    """Get icon dimensions from size attribute or icon filename.

    :param tag: Link or meta tag.
    :type tag: :class:`bs4.element.Tag`

    :return: If found, width and height, else (0,0).
    :rtype: tuple(int, int)
    """
    sizes = tag.get('sizes', '')
    if sizes and sizes != 'any':
        size = sizes.split(' ')  # '16x16 32x32 64x64'
        size.sort(reverse=True)
        width, height = re.split(r'[x\xd7/]', size[0])
    else:
        filename = tag.get('href') or tag.get('content')
        size = SIZE_RE.search(filename)
        if size:
            width, height = size.group('width'), size.group('height')
        else:
            width, height = '0', '0'

    # repair bad html attribute values: sizes='192x192+'
    width = ''.join(c for c in width if c.isdigit())
    height = ''.join(c for c in height if c.isdigit())
    return int(width), int(height)
